File: scripts/algorithm_manager.py
import re

def format_algorithm_string(alg):
    # Remove redundant spaces
    alg = re.sub(r"\s+", " ", alg).strip()
    # Capture L, L', L2, L2', etc.
    moves = re.findall(r"[RLUDFBMSExyz](?:2'?|'?2?)?", alg)
    return ' '.join(moves)

# assume input is a proper move notation
# note it can contain combinations such as R2' or R2,
# and will return R2 in both cases
def invert_move(move):
    if move.endswith("2"):
        return move  # 180-degree turns remain the same
    elif move.endswith("'"):
        return move[:-1]  # remove the apostrophe to invert
    else:
        return move + "'"  # add an apostrophe to invert

# Invert an entire algorithm
def invert_algorithm(alg):
    moves = alg.split()
    inverted_moves = [invert_move(move) for move in reversed(moves)]
    return ' '.join(inverted_moves)

def rotate_x(move):
    mapping = {'F': 'U', 'U': 'B', 'B': 'D', 'D': 'F', 'L': 'L', 'R': 'R'}
    return _transform_move(move, mapping)

def rotate_y(move):
    mapping = {'F': 'R', 'R': 'B', 'B': 'L', 'L': 'F', 'U': 'U', 'D': 'D'}
    return _transform_move(move, mapping)

def rotate_z(move):
    mapping = {'U': 'R', 'R': 'D', 'D': 'L', 'L': 'U', 'F': 'F', 'B': 'B'}
    return _transform_move(move, mapping)

def mirror_lr(move):  # swaps L<->R, keeps others
    mapping = {'L': 'R', 'R': 'L', 'U': 'U', 'D': 'D', 'F': 'F', 'B': 'B'}
    return _transform_move(move, mapping)

def mirror_fb(move):  # swaps F<->B, keeps others
    mapping = {'F': 'B', 'B': 'F', 'U': 'U', 'D': 'D', 'L': 'L', 'R': 'R'}
    return _transform_move(move, mapping)

def mirror_ud(move):  # swaps U<->D, keeps others
    mapping = {'U': 'D', 'D': 'U', 'F': 'F', 'B': 'B', 'L': 'L', 'R': 'R'}
    return _transform_move(move, mapping)

# helper function to transform a single move based on a mapping
def _transform_move(move, mapping):
    m = re.match(r"([RLUDFBMSExyz])((?:2'?)|')?", move)
    if not m:
        return move
    face, suffix = m.groups()
    if face in mapping:
        face_t = mapping[face]
    else:
        face_t = face
    # note suffix doesn't change with these transformations
    return face_t + (suffix if suffix else "")

# apply a transformation function to an entire algorithm
def transform_algorithm(alg, transform_func):
    moves = alg.strip().split()
    return ' '.join(transform_func(move) for move in moves)

File: frontend/src/utils/algorithmManager.js
// frontend/src/algorithmManager.js

// SINGLE MOVE CLASS
export class Move {
  constructor(moveStr) {
    this.move = moveStr.trim();
  }

  static from(moveStr) {
    return new Move(moveStr);
  }

  // Invert a move (R2 -> R2, R' -> R, R -> R')
  inverse() {
    if (this.move.endsWith("2")) return new Move(this.move);
    if (this.move.endsWith("'")) return new Move(this.move.slice(0, -1));
    return new Move(this.move + "'");
  }

  // For applying a rotation or mirroring function
  transform(mapping) {
    const match = this.move.match(/^([RLUDFBMSExyz])(2'?|'?2?)?$/);
    if (!match) return new Move(this.move);
    const [, face, suffix = ''] = match;
    const faceOut = mapping[face] || face;
    return new Move(faceOut + suffix);
  }

  toString() {
    return this.move;
  }
}

// ALGORITHM CLASS (LIST OF MOVES)
export class Algorithm {
  constructor(algStr) {
    // Sanitized and split
    this.moves = Algorithm._parseAlg(algStr);
  }

  static _parseAlg(algStr) {
    // Extract only valid moves
    const found = algStr.match(/[RLUDFBMSExyz](?:2'?|'?2?)?/g);
    return found ? found.map((m) => new Move(m)) : [];
  }

  static from(algStr) {
    return new Algorithm(algStr);
  }

  toString() {
    return this.moves.map((m) => m.toString()).join(" ");
  }

  inverse() {
    // Reverse and invert each move
    const inverted = this.moves.slice().reverse().map((m) => m.inverse());
    const result = new Algorithm("");
    result.moves = inverted;
    return result;
  }

  transform(transformFn) {
    // transformFn: a Move -> Move transformation
    const transformed = this.moves.map(transformFn);
    const result = new Algorithm("");
    result.moves = transformed;
    return result;
  }

  // Convenience: mirror and rotations
  rotateX() { return this.transform((m) => m.transform({F:"U",U:"B",B:"D",D:"F",L:"L",R:"R"})); }
  rotateY() { return this.transform((m) => m.transform({F:"R",R:"B",B:"L",L:"F",U:"U",D:"D"})); }
  rotateZ() { return this.transform((m) => m.transform({U:"R",R:"D",D:"L",L:"U",F:"F",B:"B"})); }
  mirrorLR() { return this.transform((m) => m.transform({L:"R",R:"L",U:"U",D:"D",F:"F",B:"B"})); }
  mirrorFB() { return this.transform((m) => m.transform({F:"B",B:"F",U:"U",D:"D",L:"L",R:"R"})); }
  mirrorUD() { return this.transform((m) => m.transform({U:"D",D:"U",F:"F",B:"B",L:"L",R:"R"})); }
}
